src/vm/amd64/CrtHelpers.asm

```diff
@@ -13,516 +13,264 @@
-        dec     r9                      ; decrement loop count
-        jnz     short mset30            ; if nz, more 8-byte blocks
-
-;
-; Fill residual bytes.
-;
-
-mset40: test    r8, r8                  ; test if any bytes to fill
-        jz      short mset60            ; if z, no bytes to fill
-mset50: mov     [rcx], dl               ; fill byte
-        inc     rcx                     ; advance to next byte
-        dec     r8                      ; decrement loop count
-        jnz     short mset50            ; if nz, more bytes to fill
-mset60: 
-        ; for some reason the assembler doesn't like the REPRET macro on the same line as a label
-        REPRET                          ; return
-
-;
-; Fill 64-byte blocks.
-;
-
-        align   16
-
-        db      066h, 066h, 066h, 090h
-        db      066h, 066h, 090h
-
-mset70: cmp     r9, CACHE_LIMIT_MEMSET / 64    ; check if large fill
-        jae     short mset90            ; if ae, large fill
-mset80: mov     [rcx], rdx              ; fill 64-byte block
-        mov     8[rcx], rdx             ;
-        mov     16[rcx], rdx            ;
-        add     rcx, 64                 ; advance to next block
-        mov     (24 - 64)[rcx], rdx     ;
-        mov     (32 - 64)[rcx], rdx     ;
-        dec     r9                      ; decrement loop count
-        mov     (40 - 64)[rcx], rdx     ;
-        mov     (48 - 64)[rcx], rdx     ;
-        mov     (56 - 64)[rcx], rdx     ;
-        jnz     short mset80            ; if nz, more 64-byte blocks
-        jmp     short mset20            ; finish in common code
-
-;
-; Fill 64-byte blocks nontemporal.
-;
-
-        align
-
-mset90: movnti  [rcx], rdx              ; fill 64-byte block
-        movnti  8[rcx], rdx             ;
-        movnti  16[rcx], rdx            ;
-        add     rcx, 64                 ; advance to next block
-        movnti  (24 - 64)[rcx], rdx     ;
-        movnti  (32 - 64)[rcx], rdx     ;
-        dec     r9                      ; decrement loop count
-        movnti  (40 - 64)[rcx], rdx     ;
-        movnti  (48 - 64)[rcx], rdx     ;
-        movnti  (56 - 64)[rcx], rdx     ;
-        jnz     short mset90            ; if nz, move 64-byte blocks
-   lock or      byte ptr [rsp], 0       ; flush data to memory
-        jmp     mset20                  ; finish in common code
+        mov     r9, 0101010101010101h   
+        imul    rdx, r9                 ; rdx is 8 bytes filler
+
+        cmp     r8, 16                  
+        jbe     mset04                 
+
+        cmp     r8, 512                 
+        jbe     mset00 
+        
+        ; count > 512
+        mov     r10, rcx                ; save dst address
+        mov     r11, rdi                ; save rdi
+        mov     eax, edx                ; eax is value
+        mov     rdi, rcx                ; rdi is dst
+        mov     rcx, r8                 ; rcx is count
+        rep     stosb
```

Comment:
Thanks @benaadams for the clarification. From that comment 
@helloguo why's the cutoff 512B? Is this based on measurement, or because of existing implementation (ex: in CRT)? Can you please add a comment about it?

Also, looks like aligning the src/dest will reduce the startup overhead for `movsb`. Is it worth trying to align the addresses befrore movsb?